src/encoding.rs

```diff
@@ -1,6 +1,21 @@
 use data_encoding::Encoding;
 use data_encoding_macro::new_encoding;
 
+// At a later point (probably when floating point arithmatic in const fn is stable) the below functions can be const
+
+#[cfg(feature = "alloc")]
+/// math comes from here https://github.com/bitcoin/bitcoin/blob/f1e2f2a85962c1664e4e55471061af0eaa798d40/src/base58.cpp#L94
+pub(crate) fn calc_encoded_size(base: usize, input_byte_size: usize) -> usize {
+    (input_byte_size as f64 * (f64::log10(256.0) / f64::log10(base as f64))) as usize + 1
+}
+
+#[cfg(feature = "alloc")]
+/// math comes from here https://github.com/bitcoin/bitcoin/blob/f1e2f2a85962c1664e4e55471061af0eaa798d40/src/base58.cpp#L48
+pub(crate) fn calc_decoded_size(base: usize, input_byte_size: usize) -> usize {
+    f64::ceil(input_byte_size as f64 * (f64::log10(base as f64) / f64::log10(256.0)) + 1.0) as usize
+    // this shouldn't need an extra one or ceiling, something is wrong somewhere
```

Comment:
Agreed. This should ideally be:

```rust
(input_byte_size as f64 * (f64::log2(base as f64) / 8.0))) as usize
// or
(input_byte_size as f64 / (8.0 / f64::log2(base as f64))) as usize
```

Reply:
I definitely agree, but for whatever reason it fails all test cases without the round up and added 1...

Just rounding up gives one short for base10 in the `preserves_two_leading_zeroes` test

```
decode len 11>11 base Base10
12
[0, 0, 121, 101, 115, 32, 109, 97, 110, 105, 32, 33]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
thread 'preserves_two_leading_zeroes' panicked at 'range end index 12 out of range for slice of length 11'
```

I will double check to make sure I am doing the tests correctly before I think about possible issues with base-x

Reply:
It's good that we have that test :) Your code is using `.len_utf8()`, that should just be `.len()` as for base encoding it's really about the bits and bytes and not about utf8 characters.